### PR TITLE
feat: trait typing

### PR DESCRIPTION
**Pull request by maartenbreddels**
Took https://github.com/ipython/traitlets/pull/788 and https://github.com/vidartf/traitlets/pull/1 did some fixes for py < 3.10 and squashed it to make rebasing to master easier.

Closes #788

### CODE REVIEW

**Comment by maartenbreddels** [comment]
Failures are expected due to not being on master (where this would be green).

Also, would require a bit more work to get all the other types supported.

As @rmorshea said in https://github.com/ipython/traitlets/pull/788#issuecomment-1291216995 there is no way but to copy-paste the `__new__` overloads and modify them for each trait. I have to say that even with more advanced typing, I don't see us avoid this, since each trait is a bit different.

**Comment by blink1073** [comment]
I'd say partial support is a good start, we can follow up with the rest.

**Comment by maartenbreddels** [comment]
I'd like to add support to a few more, and see if we can do without the `Self` and simply add some more copy-pasting.

**Comment by maartenbreddels** [comment]
My guess is that the windows failures are solved by our good friend @vidartf who is always ahead of what we need for windows: https://github.com/davidfritzsche/pytest-mypy-testing/pull/34
Is that correct @vidartf ? Does the error in CI look familiar to you?

**Comment by maartenbreddels** [comment]
cc @rmorshea 

Btw, I've moved away from overloading `__new__`, but we are now overloading `__init__`, by annotating `self`.

```python
class Float(TraitType[G, S]):
    """A float trait."""

    default_value = 0.0
    info_text = "a float"

    @t.overload
    def __init__(
        self: "Float[float, t.Union[int, float]]",
        default_value: t.Union[float, Sentinel] = ...,
        allow_none: Literal[False] = ...,
        **kwargs: t.Any,
    ):
        ...

    @t.overload
    def __init__(
        self: "Float[t.Optional[int], t.Union[int, float, None]]",
        default_value: t.Union[float, Sentinel, None] = ...,
        allow_none: Literal[True] = ...,
        **kwargs: t.Any,
    ):
        ...

    def __init__(self, default_value=Undefined, allow_none=False, **kwargs):
        self.min = kwargs.pop("min", -float("inf"))
        self.max = kwargs.pop("max", float("inf"))
        super().__init__(default_value=default_value, allow_none=allow_none, **kwargs)
```

This looks way less scary than what we suggested in #788

File: examples/docs/from_string.py [hunk 1]
```diff
@@ -25,7 +25,7 @@ class App(Application):
     )
 
     def start(self):
-        print(f"key={self.key}")
+        print(f"key={self.key.decode('utf8')}")
```
Comment (maartenbreddels) [inline examples/docs/from_string.py:28]: Typing is already paying off

File: traitlets/traitlets.py [hunk 2]
```diff
@@ -682,14 +693,63 @@ def get(self, obj, cls=None):
...
+
+    if t.TYPE_CHECKING:
+
+        @t.overload
+        def __get__(self, obj: None, cls: type[t.Any]) -> Self:
+            ...
+
+        @t.overload
+        def __get__(self, obj: t.Any, cls: type[t.Any]) -> G:
+            ...
```
[the remark this answers is not in the document]
Comment (rmorshea) [reply traitlets/traitlets.py]: This is marked outdated, but it's still a valid from what I can tell.

**Comment by rmorshea** [comment]
> I've moved away from overloading `__new__`

That makes sense to me. Users won't really care whether the descriptor is a `Float` or `Bool`, they'll just be concerned with whether the descriptor casts to and from the appropriate types when the set/get attributes (which the generic `TraitType` is fully capable of indicating). External libraries that interface with `TraitType` descriptors directly may care, but that's less common and you can convince MyPy what the narrowed type is with an `assert` check or `cast()` when you really need to.

**Comment by maartenbreddels** [comment]
> > I've moved away from overloading `__new__`
> 
> That makes sense to me. Users won't really care whether the descriptor is a `Float` or `Bool`

Overloading `__init__` this way gives you all the advantages of overloading `__new__`, except it looks more clean. So it will give you information if its either `Float` or `Bool`. 

I don't think the `__init__(self : Foo[Bar], ..` overload trick is known that well actually, I found it when looking for some issues with overloading `__new__`.

**Comment by rmorshea** [comment]
> the `__init__(self : Foo[Bar], ...)` overload trick...

Wow! I didn't even catch what was going on there. That's quite interesting. In my sample I never actually implemented a subclass so I didn't really foresee all the problems that would result from overriding `__new__` in the way I did. Having fooled around with it for a bit, my idea of copy-pasting `__new__` into subclasses does not work. For anyone else who's curious here's a [playground link](https://mypy-play.net/?mypy=latest&python=3.11&gist=09a7f574a8ec5d77cb83cdb675d5545b) that shows off this `__init__` trick in isolation.

The main downside of all this is that `self` is always unbound, that is, if you're calling a method on the parent class that's parametrized somehow, you'll need to annotate `self: TheTraitType[TheType | None]`. You can see that in the MyPy playground. If you remove the `self` annotation, you'll get an error when calling `super().__init__`.

**Comment by blink1073** [comment]
Okay, so basically we need to wait for a new release of mypy, and then if there is not a release of https://github.com/davidfritzsche/pytest-mypy-testing/pull/34, skip the typing tests on Windows.

**Comment by blink1073** [comment]
`mypy` 1.0 has been released.  It seems to me that [pytest-mypy-testing](https://github.com/davidfritzsche/pytest-mypy-testing) isn't going to get a release.  I would vote against depending on it, and instead have a set of files that we run `mypy` directly against.

**Comment by saulshanabrook** [comment]
Not to get too off-topic here, but @blink1073 I am also looking at using [pytest-mypy-testing](https://github.com/davidfritzsche/pytest-mypy-testing) in a project but also am seeing that it is out of date.

I was thinking about trying instead to use MyPy's own unit testing framework, which is built on pytest, and was curious if you all had considered that as well.

* Pytest plugin: https://github.com/python/mypy/blob/5e2702fa753ed1625d5788dcb1b71ac4a50989fd/mypy/test/data.py
* Docs on test plugin: https://github.com/python/mypy/blob/master/test-data/unit/README.md

I wasn't sure if anyone was using this outside of MyPy though.

**Comment by maartenbreddels** [comment]
Interesting, I think it is tricky to get typing right without testing both successes and failures.

**Comment by saulshanabrook** [comment]
@maartenbreddels Yes I agree. I wasn't sure if you were implying that the mypy unit testing framework doesn't support testing failures as well, but here is a simple example of using it verify some things that pass and some that fail: https://github.com/python/mypy/blob/54635dec2379e2ac8b65b6ef07778015c69cfb6a/test-data/unit/check-basic.test#L93-L99

**Comment by saulshanabrook** [comment]
I just had a go at trying to use MyPy's builtin testing framework in my package, and it seems to work OK! Only requires one monkeypatch...

https://github.com/metadsl/egg-smol-python/pull/6

**Comment by blink1073** [comment]
I wonder if the `mypy` folks would support making the plugin public and not require the monkey-patch

**Review by blink1073 — COMMENTED** [review]
lint

**Comment by blink1073** [comment]
All green, let's gooo!

**Pull request merged**